File: backend-java/src/main/java/com/peopleminer/filter/TechStackFilterService.java
package com.peopleminer.filter;

import com.peopleminer.config.TechStackConfig;
import com.peopleminer.config.TechStackConfig.RoleConfig;
import com.peopleminer.config.TechStackConfig.TargetRole;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Data;
import lombok.NoArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.stereotype.Service;

import java.util.HashSet;
import java.util.List;
import java.util.Set;
import java.util.stream.Collectors;

@Service
@Slf4j
public class TechStackFilterService {

    private final TargetRole targetRole;

    public TechStackFilterService(@Value("${app.target-role:all}") String role) {
        this.targetRole = validateRole(role);
        log.info("Tech stack filter initialized with target role: {}", this.targetRole);
    }

    private TargetRole validateRole(String role) {
        try {
            return TargetRole.valueOf(role.toUpperCase());
        } catch (IllegalArgumentException e) {
            log.warn("Invalid TARGET_ROLE \"{}\", defaulting to \"ALL\"", role);
            return TargetRole.ALL;
        }
    }

    public TargetRole getTargetRole() {
        return targetRole;
    }

    public boolean matchesTargetRole(FilterContext context) {
        return matchesRole(context, targetRole);
    }

    public boolean matchesRole(FilterContext context, TargetRole role) {
        if (role == TargetRole.ALL) {
            return true;
        }

        RoleConfig config = TechStackConfig.TECH_STACK_CONFIG.get(role);
        List<String> languages = extractLanguages(context.getRepositories());
        String textContext = buildTextContext(context);

        if (hasExcludedKeywords(textContext, config.getExcludeKeywords())) {
            return false;
        }

        if (hasExcludedLanguagesOnly(languages, config.getExcludeLanguages(), config.getLanguages())) {
            return false;
        }

        if (hasTargetLanguages(languages, config.getLanguages(), textContext, role)) {
            return true;
        }

        return hasTargetKeywords(textContext, config.getKeywords());
    }

    private List<String> extractLanguages(List<FilterContext.RepoInfo> repositories) {
        if (repositories == null) return List.of();

        return repositories.stream()
            .map(FilterContext.RepoInfo::getLanguage)
            .filter(lang -> lang != null && !lang.isEmpty())
            .distinct()
            .collect(Collectors.toList());
    }

    private String buildTextContext(FilterContext context) {
        StringBuilder sb = new StringBuilder();

        if (context.getBio() != null) {
            sb.append(context.getBio()).append(" ");
        }
        if (context.getCompany() != null) {
            sb.append(context.getCompany()).append(" ");
        }
        if (context.getRepositories() != null) {
            for (FilterContext.RepoInfo repo : context.getRepositories()) {
                if (repo.getName() != null) sb.append(repo.getName()).append(" ");
                if (repo.getDescription() != null) sb.append(repo.getDescription()).append(" ");
            }
        }

        return sb.toString().toLowerCase();
    }

    private boolean hasExcludedKeywords(String textContext, List<String> excludeKeywords) {
        if (excludeKeywords == null) return false;
        return excludeKeywords.stream()
            .anyMatch(keyword -> textContext.contains(keyword.toLowerCase()));
    }

    private boolean hasExcludedLanguagesOnly(
            List<String> languages,
            List<String> excludeLanguages,
            List<String> targetLanguages
    ) {
        if (languages.isEmpty() || excludeLanguages == null) {
            return false;
        }

        Set<String> normalizedLanguages = languages.stream()
            .map(String::toLowerCase)
            .collect(Collectors.toSet());
        Set<String> normalizedExclude = excludeLanguages.stream()
            .map(String::toLowerCase)
            .collect(Collectors.toSet());
        Set<String> normalizedTarget = targetLanguages.stream()
            .map(String::toLowerCase)
            .collect(Collectors.toSet());

        return normalizedLanguages.stream()
            .allMatch(lang -> normalizedExclude.contains(lang) && !normalizedTarget.contains(lang));
    }

    private boolean hasTargetLanguages(
            List<String> languages,
            List<String> targetLanguages,
            String textContext,
            TargetRole role
    ) {
        Set<String> normalizedLanguages = languages.stream()
            .map(String::toLowerCase)
            .collect(Collectors.toSet());
        Set<String> normalizedTarget = targetLanguages.stream()
            .map(String::toLowerCase)
            .collect(Collectors.toSet());

        for (String lang : normalizedLanguages) {
            // Handle ambiguous languages
            if (TechStackConfig.AMBIGUOUS_LANGUAGES.stream()
                    .anyMatch(al -> al.equalsIgnoreCase(lang))) {
                if (role == TargetRole.BACKEND) {
                    boolean hasBackendContext = TechStackConfig.TECH_STACK_CONFIG.get(TargetRole.BACKEND)
                        .getKeywords().stream()
                        .anyMatch(kw -> textContext.contains(kw.toLowerCase()));
                    if (hasBackendContext) return true;
                } else if (role == TargetRole.FRONTEND) {
                    boolean hasFrontendContext = TechStackConfig.TECH_STACK_CONFIG.get(TargetRole.FRONTEND)
                        .getKeywords().stream()
                        .anyMatch(kw -> textContext.contains(kw.toLowerCase()));
                    if (hasFrontendContext) return true;
                } else if (role == TargetRole.FULLSTACK) {
                    return true;
                }
                continue;
            }

            // Handle Kotlin ambiguity
            if ("kotlin".equals(lang)) {
                boolean isAndroid = TechStackConfig.KOTLIN_ANDROID_KEYWORDS.stream()
                    .anyMatch(kw -> textContext.contains(kw.toLowerCase()));
                boolean isBackend = TechStackConfig.KOTLIN_BACKEND_KEYWORDS.stream()
                    .anyMatch(kw -> textContext.contains(kw.toLowerCase()));

                if (role == TargetRole.BACKEND && isBackend && !isAndroid) return true;
                if (role == TargetRole.MOBILE && isAndroid) return true;
                if (role == TargetRole.FULLSTACK && isBackend) return true;
                continue;
            }

            if (normalizedTarget.contains(lang)) {
                return true;
            }
        }

        return false;
    }

    private boolean hasTargetKeywords(String textContext, List<String> keywords) {
        if (keywords == null) return false;
        return keywords.stream()
            .anyMatch(keyword -> textContext.contains(keyword.toLowerCase()));
    }

    public boolean matchesRoleStrict(FilterContext context, TargetRole role) {
        if (role != TargetRole.BACKEND) {
            return matchesRole(context, role);
        }

        if (!matchesRole(context, role)) {
            return false;
        }

        BackendRatioAnalysis analysis = analyzeBackendRatio(context);
        if (!analysis.isPassesFilter()) {
            log.debug("Failed backend ratio filter: {} < {}",
                analysis.getBackendRatio(), TechStackConfig.MIN_BACKEND_LANGUAGE_RATIO);
            return false;
        }

        return true;
    }

    public BackendRatioAnalysis analyzeBackendRatio(FilterContext context) {
        RoleConfig backendConfig = TechStackConfig.TECH_STACK_CONFIG.get(TargetRole.BACKEND);
        RoleConfig frontendConfig = TechStackConfig.TECH_STACK_CONFIG.get(TargetRole.FRONTEND);

        int backendCount = 0;
        int frontendCount = 0;

        Set<String> backendLangs = backendConfig.getLanguages().stream()
            .map(String::toLowerCase)
            .collect(Collectors.toSet());
        Set<String> frontendLangs = frontendConfig.getLanguages().stream()
            .map(String::toLowerCase)
            .collect(Collectors.toSet());
        Set<String> ambiguous = TechStackConfig.AMBIGUOUS_LANGUAGES.stream()
            .map(String::toLowerCase)
            .collect(Collectors.toSet());

        for (FilterContext.RepoInfo repo : context.getRepositories()) {
            if (repo.getLanguage() == null) continue;
            String lang = repo.getLanguage().toLowerCase();

            if (ambiguous.contains(lang)) {
                String repoContext = ((repo.getName() != null ? repo.getName() : "") + " " +
                    (repo.getDescription() != null ? repo.getDescription() : "")).toLowerCase();

                boolean hasBackendContext = backendConfig.getKeywords().stream()
                    .anyMatch(kw -> repoContext.contains(kw.toLowerCase()));
                boolean hasFrontendContext = frontendConfig.getKeywords().stream()
                    .anyMatch(kw -> repoContext.contains(kw.toLowerCase()));

                if (hasBackendContext && !hasFrontendContext) backendCount++;
                else if (hasFrontendContext && !hasBackendContext) frontendCount++;
                continue;
            }

            if ("kotlin".equals(lang)) {
                String repoContext = ((repo.getName() != null ? repo.getName() : "") + " " +
                    (repo.getDescription() != null ? repo.getDescription() : "")).toLowerCase();

                boolean isAndroid = TechStackConfig.KOTLIN_ANDROID_KEYWORDS.stream()
                    .anyMatch(kw -> repoContext.contains(kw.toLowerCase()));
                boolean isBackend = TechStackConfig.KOTLIN_BACKEND_KEYWORDS.stream()
                    .anyMatch(kw -> repoContext.contains(kw.toLowerCase()));

                if (isBackend && !isAndroid) backendCount++;
                continue;
            }

            if (backendLangs.contains(lang)) backendCount++;
            else if (frontendLangs.contains(lang)) frontendCount++;
        }

        int total = backendCount + frontendCount;
        double backendRatio = total > 0 ? (double) backendCount / total : 0;

        return BackendRatioAnalysis.builder()
            .backendCount(backendCount)
            .frontendCount(frontendCount)
            .backendRatio(backendRatio)
            .passesFilter(backendRatio >= TechStackConfig.MIN_BACKEND_LANGUAGE_RATIO)
            .build();
    }

    @Data
    @NoArgsConstructor
    @AllArgsConstructor
    @Builder
    public static class FilterContext {
        private List<RepoInfo> repositories;
        private String bio;
        private String company;

        @Data
        @NoArgsConstructor
        @AllArgsConstructor
        @Builder
        public static class RepoInfo {
            private String language;
            private String name;
            private String description;
        }
    }

    @Data
    @NoArgsConstructor
    @AllArgsConstructor
    @Builder
    public static class BackendRatioAnalysis {
        private int backendCount;
        private int frontendCount;
        private double backendRatio;
        private boolean passesFilter;
    }
}

File: backend/src/filter/tech-stack-filter.service.ts
import { Injectable, Logger } from '@nestjs/common';
import { ConfigService } from '@nestjs/config';
import {
  TECH_STACK_CONFIG,
  TargetRole,
  AMBIGUOUS_LANGUAGES,
  KOTLIN_ANDROID_KEYWORDS,
  KOTLIN_BACKEND_KEYWORDS,
  BACKEND_FILTER_CONFIG,
} from '../config/tech-stack.config';

interface Repository {
  language: string | null;
  name?: string;
  description?: string | null;
}

export interface FilterContext {
  repositories: Repository[];
  bio?: string | null;
  company?: string | null;
}

@Injectable()
export class TechStackFilterService {
  private readonly logger = new Logger(TechStackFilterService.name);
  private readonly targetRole: TargetRole;

  constructor(private configService: ConfigService) {
    const role = this.configService.get<string>('TARGET_ROLE', 'all');
    this.targetRole = this.validateRole(role);
    this.logger.log(`Tech stack filter initialized with target role: ${this.targetRole}`);
  }

  private validateRole(role: string): TargetRole {
    const validRoles: TargetRole[] = ['backend', 'frontend', 'mobile', 'fullstack', 'all'];
    if (validRoles.includes(role as TargetRole)) {
      return role as TargetRole;
    }
    this.logger.warn(`Invalid TARGET_ROLE "${role}", defaulting to "all"`);
    return 'all';
  }

  getTargetRole(): TargetRole {
    return this.targetRole;
  }

  matchesTargetRole(context: FilterContext): boolean {
    return this.matchesRole(context, this.targetRole);
  }

  matchesRole(context: FilterContext, role: TargetRole): boolean {
    if (role === 'all') {
      return true;
    }

    const config = TECH_STACK_CONFIG[role];
    const languages = this.extractLanguages(context.repositories);
    const textContext = this.buildTextContext(context);

    if (this.hasExcludedKeywords(textContext, config.excludeKeywords)) {
      return false;
    }

    if (this.hasExcludedLanguagesOnly(languages, config.excludeLanguages, config.languages)) {
      return false;
    }

    if (this.hasTargetLanguages(languages, config.languages, textContext, role)) {
      return true;
    }

    if (this.hasTargetKeywords(textContext, config.keywords)) {
      return true;
    }

    return false;
  }

  private extractLanguages(repositories: Repository[]): string[] {
    const languages = repositories
      .map((repo) => repo.language)
      .filter((lang): lang is string => lang !== null);
    return [...new Set(languages)];
  }

  private buildTextContext(context: FilterContext): string {
    const parts: string[] = [];

    if (context.bio) {
      parts.push(context.bio);
    }

    if (context.company) {
      parts.push(context.company);
    }

    context.repositories.forEach((repo) => {
      if (repo.name) parts.push(repo.name);
      if (repo.description) parts.push(repo.description);
    });

    return parts.join(' ').toLowerCase();
  }

  private hasExcludedKeywords(textContext: string, excludeKeywords: string[]): boolean {
    return excludeKeywords.some((keyword) => textContext.includes(keyword.toLowerCase()));
  }

  private hasExcludedLanguagesOnly(
    languages: string[],
    excludeLanguages: string[],
    targetLanguages: string[],
  ): boolean {
    if (languages.length === 0) {
      return false;
    }

    const normalizedLanguages = languages.map((l) => l.toLowerCase());
    const normalizedExclude = excludeLanguages.map((l) => l.toLowerCase());
    const normalizedTarget = targetLanguages.map((l) => l.toLowerCase());

    const hasOnlyExcludedLanguages = normalizedLanguages.every(
      (lang) =>
        normalizedExclude.includes(lang) && !normalizedTarget.includes(lang),
    );

    return hasOnlyExcludedLanguages;
  }

  private hasTargetLanguages(
    languages: string[],
    targetLanguages: string[],
    textContext: string,
    role: TargetRole,
  ): boolean {
    const normalizedLanguages = languages.map((l) => l.toLowerCase());
    const normalizedTarget = targetLanguages.map((l) => l.toLowerCase());

    for (const lang of normalizedLanguages) {
      if (AMBIGUOUS_LANGUAGES.map((l) => l.toLowerCase()).includes(lang)) {
        if (role === 'backend') {
          const hasBackendContext = TECH_STACK_CONFIG.backend.keywords.some((kw) =>
            textContext.includes(kw.toLowerCase()),
          );
          if (hasBackendContext) return true;
        } else if (role === 'frontend') {
          const hasFrontendContext = TECH_STACK_CONFIG.frontend.keywords.some((kw) =>
            textContext.includes(kw.toLowerCase()),
          );
          if (hasFrontendContext) return true;
        } else if (role === 'fullstack') {
          return true;
        }
        continue;
      }

      if (lang === 'kotlin') {
        const isAndroid = KOTLIN_ANDROID_KEYWORDS.some((kw) =>
          textContext.includes(kw.toLowerCase()),
        );
        const isBackend = KOTLIN_BACKEND_KEYWORDS.some((kw) =>
          textContext.includes(kw.toLowerCase()),
        );

        if (role === 'backend' && isBackend && !isAndroid) return true;
        if (role === 'mobile' && isAndroid) return true;
        if (role === 'fullstack' && isBackend) return true;
        continue;
      }

      if (normalizedTarget.includes(lang)) {
        return true;
      }
    }

    return false;
  }

  private hasTargetKeywords(textContext: string, keywords: string[]): boolean {
    return keywords.some((keyword) => textContext.includes(keyword.toLowerCase()));
  }

  detectRole(context: FilterContext): TargetRole {
    const roles: TargetRole[] = ['backend', 'frontend', 'mobile'];
    const matches: TargetRole[] = [];

    for (const role of roles) {
      if (this.matchesRole(context, role)) {
        matches.push(role);
      }
    }

    if (matches.length === 0) return 'all';
    if (matches.length === 1) return matches[0];
    if (matches.includes('backend') && matches.includes('frontend')) return 'fullstack';

    return matches[0];
  }

  /**
   * Analyze backend language ratio for strict filtering
   */
  analyzeBackendRatio(context: FilterContext): {
    backendCount: number;
    frontendCount: number;
    backendRatio: number;
    passesFilter: boolean;
  } {
    const backendConfig = TECH_STACK_CONFIG.backend;
    const frontendConfig = TECH_STACK_CONFIG.frontend;

    let backendCount = 0;
    let frontendCount = 0;

    for (const repo of context.repositories) {
      if (!repo.language) continue;

      const lang = repo.language.toLowerCase();
      const normalizedBackendLangs = backendConfig.languages.map((l) =>
        l.toLowerCase(),
      );
      const normalizedFrontendLangs = frontendConfig.languages.map((l) =>
        l.toLowerCase(),
      );

      // Check if it's an ambiguous language (TypeScript/JavaScript)
      if (
        AMBIGUOUS_LANGUAGES.map((l) => l.toLowerCase()).includes(lang)
      ) {
        const repoContext = `${repo.name || ''} ${repo.description || ''}`.toLowerCase();
        const hasBackendContext = backendConfig.keywords.some((kw) =>
          repoContext.includes(kw.toLowerCase()),
        );
        const hasFrontendContext = frontendConfig.keywords.some((kw) =>
          repoContext.includes(kw.toLowerCase()),
        );

        if (hasBackendContext && !hasFrontendContext) {
          backendCount++;
        } else if (hasFrontendContext && !hasBackendContext) {
          frontendCount++;
        }
        // If both or neither, don't count toward ratio
        continue;
      }

      // Check Kotlin ambiguity
      if (lang === 'kotlin') {
        const repoContext = `${repo.name || ''} ${repo.description || ''}`.toLowerCase();
        const isAndroid = KOTLIN_ANDROID_KEYWORDS.some((kw) =>
          repoContext.includes(kw.toLowerCase()),
        );
        const isBackend = KOTLIN_BACKEND_KEYWORDS.some((kw) =>
          repoContext.includes(kw.toLowerCase()),
        );

        if (isBackend && !isAndroid) {
          backendCount++;
        }
        // Android Kotlin is neither backend nor frontend for this calculation
        continue;
      }

      // Direct language matching
      if (normalizedBackendLangs.includes(lang)) {
        backendCount++;
      } else if (normalizedFrontendLangs.includes(lang)) {
        frontendCount++;
      }
    }

    const total = backendCount + frontendCount;
    const backendRatio = total > 0 ? backendCount / total : 0;

    return {
      backendCount,
      frontendCount,
      backendRatio,
      passesFilter: backendRatio >= BACKEND_FILTER_CONFIG.minBackendLanguageRatio,
    };
  }

  /**
   * Strict role matching with ratio filter for backend
   */
  matchesRoleStrict(context: FilterContext, role: TargetRole): boolean {
    // For non-backend roles, use standard matching
    if (role !== 'backend') {
      return this.matchesRole(context, role);
    }

    // First check basic matching
    if (!this.matchesRole(context, role)) {
      return false;
    }

    // Apply ratio filter for backend
    const ratioAnalysis = this.analyzeBackendRatio(context);

    if (!ratioAnalysis.passesFilter) {
      this.logger.debug(
        `Failed backend ratio filter: ${ratioAnalysis.backendRatio.toFixed(2)} < ${BACKEND_FILTER_CONFIG.minBackendLanguageRatio}`,
      );
      return false;
    }

    return true;
  }
}
